utils/heap/map.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package heap
+
+import (
+	"container/heap"
+
+	"github.com/ava-labs/avalanchego/utils"
+)
+
+var _ heap.Interface = (*indexedQueue[int, int])(nil)
+
+func MapToList[K comparable, V any](m Map[K, V]) []V {
+	result := make([]V, 0, m.Len())
+
+	for _, e := range m.queue.entries {
+		result = append(result, e.v)
+	}
+
+	return result
+}
+
+// NewMap returns a heap without duplicates ordered by its values
+func NewMap[K comparable, V any](less func(a, b V) bool) Map[K, V] {
+	return Map[K, V]{
+		queue: &indexedQueue[K, V]{
+			queue: queue[entry[K, V]]{
+				entries: make([]entry[K, V], 0),
+				less: func(a, b entry[K, V]) bool {
+					return less(a.v, b.v)
+				},
+			},
+			index: make(map[K]int),
+		},
+	}
+}
+
+type Map[K comparable, V any] struct {
+	queue *indexedQueue[K, V]
+}
+
+// Push returns the evicted previous value if present
+func (m *Map[K, V]) Push(k K, v V) (V, bool) {
+	if i, ok := m.queue.index[k]; ok {
+		prev := m.queue.entries[i]
+		m.queue.entries[i].v = v
+		heap.Fix(m.queue, i)
+		return prev.v, true
+	}
+
+	heap.Push(m.queue, entry[K, V]{k: k, v: v})
+	return utils.Zero[V](), false
+}
+
+func (m *Map[K, V]) Pop() (K, V, bool) {
+	if m.Len() == 0 {
+		return utils.Zero[K](), utils.Zero[V](), false
+	}
+
+	popped := heap.Pop(m.queue).(entry[K, V])
+	return popped.k, popped.v, true
+}
+
+func (m *Map[K, V]) Peek() (K, V, bool) {
+	if m.Len() == 0 {
+		return utils.Zero[K](), utils.Zero[V](), false
+	}
+
+	entry := m.queue.entries[0]
+	return entry.k, entry.v, true
+}
+
+func (m *Map[K, V]) Len() int {
+	return m.queue.Len()
+}
+
+func (m *Map[K, V]) Remove(k K) (K, V, bool) {
```

Comment:
do we really need to return the removed key here? Consider returning only value and boolean

Reply:
Yeah we can remove this